### ISSUE TITLE
CollectionFS

### ISSUE THREAD

**Issue by cyrus-za** [Urigo/angular-meteor#72]
I am not sure if I am doing it right, but how do you use angular-meteor with collectionFS?

If I try to bind to a collectionFS object (using gridFS for storage) it throws an error that the first parameter needs to be a meteor.Collection.

I declare my model as such: 

    SomeCollection = new FS.Collection('SomeCollection', {
      stores: [new FS.Store.GridFS("someCollection")]
    });

in my controller I add this line: 

    angular.module('myApp').controller('myCtrl', function($scope, $collection){
       $collection(SomeCollection).bind($scope,'someCollection',true,true);
    });

**Comment by epotvin** [comment]
I have some issues too with CollectionFS. I resolved the collection problem with a function :

```javascript
$scope.images = $meteorCollection(function() {
    return Images.find();
}).subscribe('images');
```
I even find a way to get the image url with a method :

```javascript
$scope.url = function(image) {
    return Images.findOne(image._id).url();
};
```
But I still have issues with removing...
`$scope.images.remove(image);` doesn't work at all... And a plain `Images.remove(image._id)` works but Angular is complaining...

**Comment by Urigo** [comment]
@epotvin that's a cool solution.
Can you upload an example repo so I can look at your problems?
then I could add this to the docs and the tutorial

**Comment by Urigo** [comment]
Would you think it's a good idea to add $meteor.collectionFS service?
If so, anyone wants to start thinking about how will it look like?

**Comment by netanelgilad** [comment]
OK so here is what I think we need to do here, ideas welcome.
First change the way we handle getting the the cursor from a collection. Instead of checking if the object we got is a `Mongo.Collection`, just check that it has a `find` method and that the return value from the `find` method is a `Mongo.Cursor`. Then, that should allow to bind to the `FS.Collection` object since it has a `find` method and it returns a `Mongo.Cursor`. The other methods needed for the `$meteor.collection` to work are `insert` and `remove` and they should exist on the `FS.Collection`.
Then if someone wants to work with a cursor from the `FS.Collection`, then we need to add a way to send the collection object as well as the cursor since we can't get the collection from the cursor the same way we do today. That could be done by adding it as an argument to `$meteor.collection`.
Thoughts?

**Comment by Urigo** [comment]
Can you branch and show an example of your thoughts?
might be nicer to talk seeing code (doesn't need to actually work at the beginning)

**Comment by netanelgilad** [comment]
@Urigo check out the changes in the commit. Haven't checked it out yet but that is the simple idea and it should make it work with collectionFS or any other object that is collection-like.
If any of the guys in this thread have a way I can test this code against their use case it would be great and would save me the need to write another sample project =]

**Comment by epotvin** [comment]
I guys. I managed to create a really simple example of this problem here :
https://github.com/epotvin/thumbnail-example

I'm still working on it and I will try the change of @netanelgilad as soon as I am done. But as you will see, I have some more issues I'm trying to fix.

Keep me posted and don't hesitate if you have anything you want me to try or explain further about what I want to achieve here. Anyway I consider the stuff I'm able to do with what I have right now is a lot better then anything availble.

**Comment by netanelgilad** [comment]
So I checked your example @epotvin, and it works with the updates in my branch!
@Urigo There was just a small problem. There was an `angular.copy` of the object being inserted/updated and that causes problems when the object being inserted is something like a `File` object, which angular doesn't copy that well. Look at this commit https://github.com/Urigo/angular-meteor/commit/9d857e5159aa56d80736b2fe3a873c2ec54a3f9b.
Do you know why it was there and what it is needed for?

**Comment by epotvin** [comment]
Thanks @netanelgilad ! I just tried too but I wasn't able to make it works as is since you branched from an older version of the repo. I suppose I had been able to fix it by renaming my templates with the .tpl extension, but instead I made a bold try and merge back master in your branch (https://github.com/Kogitec/angular-meteor/commit/6ea88d39409ec76290bbf3f7c649215795908654)

I had no conflict but when I want to instantiate a collection by passing only the collection name, as in 

```javascript
    $scope.images = $meteor.collection(Images);
```
CollectionFS is complaining that the find function is undefined. It's because in FS.Collection.find, the this variable points to Window instead of the actual instance. (https://github.com/CollectionFS/Meteor-CollectionFS/blob/master/packages/collection/api.common.js#L174)

Still working on it.

**Comment by Urigo** [comment]
@epotvin have you tried?

    $scope.images = $meteor.collection(Images, false);

**Comment by netanelgilad** [comment]
Did you tried against the latest commit in branch? There is a commit that fixes it

**Comment by epotvin** [comment]
Oh yes it's working! With of without the second parameter set to false.

I guess the same fix could be applied to the object function. Actually, the following line still complains about the collection type

```javascript
$scope.image = $meteor.object(Images, $stateParams.imageId);
```

**Comment by Urigo** [comment]
Fixed on 0.8.3

@epotvin that's a really good idea about $meteor.object, can you open that in a separate issue?

Closing the longest running issue we have!!

**Comment by epotvin** [comment]
Yes I will, and I'll try to fix it, supposing I could replicate the logic you used with the collection.
